### DISCUSSION TITLE
Clipping Selection

### DISCUSSION THREAD

**Discussion by hanan-commits-work** [iTwin/itwinjs-core#5181] [Q&A] [answered]
Greetings.
I managed to do clipping for discrete shape following the Clipping of Planeset in the new iTwins 3.5 feature.
But I am unable to deselect the element clipped. 
See the [Sandbox](https://www.itwinjs.org/sandboxes/StefanRetief/View%20Clipping%20Sample). Ideally, the highlight should disappear. and the element clipped should appear "unselected"

cc: @pmconne

**Comment by pmconne** [comment, marked as answer]
I'm sorry, I need you to explain what actions I need to perform in your sandbox and what the expected vs observed results should be.

**Reply by hanan-commits-work** [reply]
Hey Paul.
If you take a look at this sandbox, it shows the ideal behavior. Select an element. It'd get highlighted. But when you click elsewhere on the screen, the element gets deselected. 

Now what is happening in my project, I select an element, and perform clipping based on the new iTwins 3.5 features, the element appears selected (highlighted). When I click elsewhere, the clipped element should get deselected and no longer highlighted. But that is not the case in my app.

Let me know if you have any questions?

**Reply by pmconne** [reply]
Sorry, are you saying the sandbox you linked to does not exhibit the behavior you are describing?
If so, provide a link to a sandbox that does exhibit the behavior. Or at the very least, show your code, so that we can see what you mean by "perform clipping based on the new iTwins 3.5 features".

**Reply by hanan-commits-work** [reply]
hey @pmconne. Apologies for the confusion. Let me walk you through the current situation:

1. I have selected an element in order to perform planar clip masking (Highlighted element):
![image](https://user-images.githubusercontent.com/89520756/227121895-45b185a9-b894-46e5-89da-e6a6faa45898.png)

2. Now, when I perform masking, I get this result.

![image](https://user-images.githubusercontent.com/89520756/227122041-9ad563d8-0a78-4cf4-81d9-62a025fbcfc9.png)

3. Basically, I want to remove the highlighted to "Isolate" the element. But when, I try to remove the highlight by clicking somewhere else on the screen, I get this:

![image](https://user-images.githubusercontent.com/89520756/227122260-7d7983c1-c78f-429c-a00e-ffbdc792bdb4.png)

^Please note that highlight is "removed" but the white lines in cylinder dont disappear.


I am using the following code:
```

  /* Method for adding decorators to the viewport */
  public static addDecorators(vp: ScreenViewport) {
    // Create a clip decorator. Selecting the clip decoration to immediately show the handles is the default.
    const vcdp: ViewClipDecorationProvider = ViewClipDecorationProvider.create();
    // Default behavior is to hide the decorators on deselect. We want to keep the decorators showing in this example.
    vcdp.clearDecorationOnDeselect = false;
    vcdp.showDecoration(vp);
    // The decorators require the SelectTool being active.
    IModelApp.toolAdmin.startDefaultTool();
  }

  public static setViewClipFromClipPlaneSet(vp: ScreenViewport, planeSet: ConvexClipPlaneSet) {
    const prim = ClipPrimitive.createCapture(planeSet);
    const clip = ClipVector.createEmpty();
    clip.appendReference(prim);
    vp.view.setViewClip(clip);
    vp.setupFromView();
    this.addDecorators(vp);
    //this.isolateSelectedElements(vp);
    return true;
  }

  public static async attachMask(maskModel: ContextRealityModelState, maskElem: Set<string>) {   
    if (maskElem.size !== 0) {
      const modelIds = await cView.fetchModelIds("$Elements"); //We get the modelIds of the selected Elements here
      const elementIterator = maskElem.values();
      const elementId64 = elementIterator.next().value;
      const elementMesh: ElementMeshRequestProps = {source: elementId64};
      const polyface = await this.iModel.generateElementMeshes(elementMesh);
      const decodedPolyfaces = readElementMeshes(polyface);
      const convexPlaneSet = ConvexClipPlaneSet.createConvexPolyface(decodedPolyfaces[0]);
      return this.setViewClipFromClipPlaneSet(this.sViewPort, convexPlaneSet.clipper);
  }
   else {
     cMessage.showStatusBarError("Element Not Selected");
    }
  }

```

Hope this provides more clarity. Let me know if there are any questions

**Reply by pmconne** [reply]
That looks like the clip decorations, not the element. Comment out your call to `addDecorators`.

**Reply by hanan-commits-work** [reply]
Hi @pmconne. Thank you for all your help. The problem was indeed in clip decorations. 

There's another issue I would like your help in. I am trying to clip multiple elements. To do this, I created an iterator in the attachMask method above. But instead of all the elements being clipped, it clips **one element** in the selection, removs the clipping from the first element, and clips the next. 

Following is the code im using.
```

  public static async attachMask(maskModel: ContextRealityModelState, maskElem: Set<string>) {   
    if (maskElem.size !== 0) {
      const modelIds = await cView.fetchModelIds("$Elements");
      const elementIterator = maskElem.values(); // getting array
      const polyfaces: any = [];
      for (const elementId64 of elementIterator) {
        const elementMesh: ElementMeshRequestProps = { source: elementId64 };
        const polyface = await this.iModel.generateElementMeshes(elementMesh);
        console.log(polyface, elementId64, elementMesh);
        const decodedPolyfaces = readElementMeshes(polyface);
        const convexPlaneSet = ConvexClipPlaneSet.createConvexPolyface(decodedPolyfaces[0]);
        this.setViewClipFromClipPlaneSet(this.sViewPort, convexPlaneSet.clipper);
      }
      cMessage.showInfo("Clipping started")
      return;
  }
   else {
     cMessage.showStatusBarError("Element Not Selected");
    }
  }
```

**Reply by pmconne** [reply]
Your code as written throws away all but the last element's clips. You need to create a [UnionOfConvexClipPlaneSets](https://www.itwinjs.org/reference/core-geometry/cartesiangeometry/unionofconvexclipplanesets/) combining the meshes of each element.

Create a ConvexClipPlaneSet from each element as you are currently doing and add it to a list. After your loop, use `UnionOfConvexClipPlaneSets.createConvexSets` to union all of your clip plane sets. Pass the union to [ClipPrimitive.createCapture](https://www.itwinjs.org/reference/core-geometry/cartesiangeometry/clipprimitive/createcapturestatic/), then create a ClipVector from the ClipPrimitive and apply it to your view.

**Reply by hanan-commits-work** [reply]
Just one clarification. Do I need to use ClipVector.create or ClipVector.createCapture method? 
I am getting type errors in both...

**Reply by pmconne** [reply]
Consult the documentation. If problems persist, post your code and the errors.

**Reply by hanan-commits-work** [reply]
The problem is persisting. I am getting the following error:

![image](https://user-images.githubusercontent.com/89520756/228302342-2b824f00-37d7-4c36-9e92-ce60225713fe.png)

This is the my code:
```
 const elementIterator = maskElem.values();
      const convexPlaneSetsList: any = [];
      for (const elementId64 of elementIterator) {
        const elementMesh: ElementMeshRequestProps = { source: elementId64 };
        const polyface = await this.iModel.generateElementMeshes(elementMesh);
        const decodedPolyfaces = readElementMeshes(polyface);
        const convexPlaneSet = ConvexClipPlaneSet.createConvexPolyface(decodedPolyfaces[0]);
        convexPlaneSetsList.push(convexPlaneSet);
      }
      const unionConvexClipPlaneSets = UnionOfConvexClipPlaneSets.createConvexSets(convexPlaneSetsList);
      const clipPrimitiveList: any = [];
      clipPrimitiveList.push(ClipPrimitive.createCapture(unionConvexClipPlaneSets));
      const clipPrimitive = ClipPrimitive.createCapture(unionConvexClipPlaneSets);
      const clipVector: ClipVector = ClipVector.createEmpty();
      clipVector.appendReference(clipPrimitive);
      this.sViewPort.view.setViewClip(clipVector);  // Error is occurring at this line
      return;
```

**Comment by pmconne** [comment]
> `const convexPlaneSetsList: any`

If you subvert the type system, the compiler can't tell you what you're doing wrong.
Look at the return type of `createConvexPolyface`, and the way you were using it in your previous code.

**Reply by ovaisa** [reply]
We are able to get this working as intended but we are experiencing performance issues when we have more than 4-5 elements.  I believe it is because our classifier model consists of pipes (cylinders) and elbows (torus) which is generating a large number of faces.  We are planning to simplify the classifier model and use an extruded rectangle or a cube (solid) in MicroStation.  

I wanted to know if using a certain type of primitives (solid vs. extruded vs ???) would be more efficient for this purpose (of creating ClipVector).  I will advise our CAD modelers accordingly.

Thanks,

Ovais

**Reply by pmconne** [reply]
Cube vs extruded rectangle shouldn't make a difference in the number of faces produced for the clip vector.
